File: Clean-Architecture/Feature/dart/example/test/features/auth/domain/usecases/login_user_test.dart
import 'package:clean_architecture_feature_based_flutter_example/core/result/result.dart';
import 'package:clean_architecture_feature_based_flutter_example/features/auth/domain/entities/auth_session.dart';
import 'package:clean_architecture_feature_based_flutter_example/features/auth/domain/entities/user.dart';
import 'package:clean_architecture_feature_based_flutter_example/features/auth/domain/repositories/auth_repository.dart';
import 'package:clean_architecture_feature_based_flutter_example/features/auth/domain/usecases/login_user.dart';
import 'package:flutter_test/flutter_test.dart';

class FakeAuthRepository implements AuthRepository {
  String? capturedEmail;

  @override
  Future<Result<AuthSession>> login({required String email, required String password}) async {
    capturedEmail = email;
    return const Success(AuthSession(
      user: User(id: '1', name: 'Anwar', email: 'anwar@example.com'),
      token: 'token',
    ));
  }
}

void main() {
  test('LoginUser normalizes email and delegates to repository', () async {
    final repository = FakeAuthRepository();
    final useCase = LoginUser(repository);
    final result = await useCase(
      email: ' ANWAR@EXAMPLE.COM ',
      password: 'password123',
    );

    expect(repository.capturedEmail, 'anwar@example.com');
    expect(result, isA<Success<AuthSession>>());
  });
}

File: Clean-Architecture/Feature/Javascript/example/test/auth/login-user.test.js
import test from 'node:test';
import assert from 'node:assert/strict';
import { LoginUser } from '../../src/features/auth/domain/usecases/LoginUser.js';

class FakeAuthRepository {
  async login({ email, password }) {
    if (email === 'anwar@example.com' && password === 'password123') {
      return { user: { id: '1', email }, token: 'token' };
    }
    return null;
  }
}

test('LoginUser returns a user and token for valid credentials', async () => {
  const useCase = new LoginUser({ authRepository: new FakeAuthRepository() });
  const result = await useCase.execute({
    email: ' ANWAR@EXAMPLE.COM ',
    password: 'password123',
  });
  assert.equal(result.user.email, 'anwar@example.com');
  assert.equal(result.token, 'token');
});

test('LoginUser rejects invalid credentials', async () => {
  const useCase = new LoginUser({ authRepository: new FakeAuthRepository() });
  await assert.rejects(
    () => useCase.execute({ email: 'wrong@example.com', password: 'bad' }),
    (error) => error.code === 'INVALID_CREDENTIALS' && error.statusCode === 401,
  );
});
